feat: 캐싱을 활용한 인증 로직 개선 및 성능 최적화

jooyeon0907 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**  
- `getCalendar()`, `getSchedule()` 메서드에 캐싱 처리가 되어있지 않음.  

**TO-BE**  
1. **Redis 캐시 설정**  
   - 라이브러리 추가:
     - `spring-boot-starter-cache`
     - `jackson-datatype-jsr310`
   - `@Cacheable` 어노테이션을 사용하여 캐시 저장.
   - 수정 요청 시 `RedisTemplate`을 사용하여 캐시를 업데이트:
     - `@Cacheable` 메서드의 반환 값은 `Calendar` / `Schedule` 객체.
     - 수정 요청 메서드의 반환 값은 `ResponseDto`이므로 `RedisTemplate`을 사용.
   - 삭제 요청 메서드에 `@CacheEvict` 추가:
     - 삭제 성공 시, 해당 캐시를 제거.
     - 단, 캐시 키가 사용자 ID + 캘린더 ID 조합이기 때문에 다른 사용자 ID의 캐시는 삭제되지 않음.
     - `redisTemplate.keys()`로 키 검색 및 삭제는 비효율적이므로, 캐시 유효시간(TTL)을 30분으로 설정하여 자동 삭제되도록 처리.

2. **Redis 설정**
   - `cacheManager` 설정: `@Cacheable` 사용 시 캐시 처리.
   - `redisTemplate` 설정: 수정 요청 시 직접 캐시 업데이트를 위해 사용.
   - 직렬화 설정: `GenericJackson2JsonRedisSerializer` 사용.
   - 캐시 TTL 설정: 30분.

3. **서비스 메서드 이름 변경**
   - `check~` 메서드를 `validate~`로 이름 변경.

4. **Schedule 엔티티 수정**
   - `Schedule` 엔티티의 `calendar_id` 필드에 누락된 cascade 설정 추가 (캘린더 삭제 시 연관된 일정 데이터 삭제).

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드  
  - **캐시 읽기 테스트 (`testCalendarCaching_read`)**:  
    - `commonService.getCalendar()` 메서드를 두 번 호출하여 동일한 결과 반환 확인.  
    - 두 번째 호출은 캐시에서 값을 가져와야 하며, DB 조회 없이 동일한 값 반환.  
    - `CacheManager`를 사용하여 캐시 저장소의 값 확인.  
  - **캐시 조회 성능 테스트 (`testCacheHitPerformance`)**:  
    - 첫 번째 호출은 캐시 미스로 인해 DB 조회.  
    - 두 번째 호출은 캐시 히트로 인해 빠르게 데이터 반환.  
    - 호출 시간 비교를 통해 캐시 성능 향상 검증.  
  - **캐시 업데이트 테스트 (`testCalendarCaching_update`)**:  
    - `commonService.getCalendar()` 호출 후, `calendarService.update()`를 사용하여 데이터 수정.  
    - 캐시 값도 함께 갱신되는지 확인하기 위해 수정된 값과 비교.  
    - `CacheManager`로 캐시를 다시 조회하여 수정된 값 반영 여부 확인.  
- [x] Redis에 값이 저장 및 삭제되는 것 확인 완료.

jooyeon0907 committed:
feat: 캐싱을 활용한 인증 로직 개선 및 성능 최적화

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarService.java:
```diff
@@ -51,36 +52,34 @@ public List<CalendarDto.Response> list(Long userId) {
 		calendar.setName(request.getName());
 		calendarRepository.save(calendar);
 
+		// @Cacheable 로 캐시 저장할 때와 반환 값이 달라서 redisTemplate 사용해 캐시 업데이트
```
반환값이 달라서 @Cacheable 을 쓰지 못한다는 것인가요?

mingming-mentor replied:
제가 주석을 제대로 잘 이해못했는데 만약 적어주신 내용이, method가 update 메소드라서 처음 저장시와 달라 캐시 업데이트가 필요하여 redis에 직접 넣어주시는 거라면 @CachePut 을 이용하시면 좋을 것 같아요.
캐시 관련해서 어노테이션을 쓰셔서 코드 관리하는 것으로 방향을 잡으셨다면 일관되게 어노테이션으로 관리하면 좋을 것 같아요

jooyeon0907 replied:
아 반환 타입이 다르다는 의미였습니다! 반환값이라 적어서 헷갈렸을 수 있겠네요..!
update 처리하는 메소드 반환 타입은 CalendarDto.Response 이고, getCalendar()  반환 타입은 Calendar 이여서 
@CachePut 시 캐싱된 데이터에서 @class 프로퍼티 값이 달라져서 오류가 발생하여
Calendar 객체로 캐싱되도록 하려고 redisTemplate 을 사용했던 것이었습니다.

그런데 entity를 그대로 캐싱하시면 안된다는 리뷰를 반영하여
CalendarDto.Response 로 캐싱하도록 하고, redisTemplate 사용해 캐시 업데이트 하는 코드 삭제 및 어노테이션으로 캐시 관리하도록 수정하겠습니다!

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarService.java:
```diff
@@ -28,6 +27,8 @@ public class CalendarService {
 	private final CalendarRepository calendarRepository;
 	private final UserCalendarRepository userCalendarRepository;
 
+	private final RedisTemplate<String, Object> redisTemplate;
```
calendar 저장용 타입을 별도로 만들고 RedisTemplate<String, CalenderDto> 로 받으시면 좋겠네요

jooyeon0907 replied:
어노테이션으로 캐시 관리하도록 통일 시킬 것이어서 저 부분은 필요 없어져 삭제하겠습니다! 
만약 redisTemplate 사용하여 캐시하게 된다면 Object 가 아닌 CalenderDto 로 받는 게 좋다는 점 참고하겠습니다 감사합니다

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarService.java:
```diff
@@ -51,36 +52,34 @@ public List<CalendarDto.Response> list(Long userId) {
 		calendarRepository.save(calendar);
 
+		// @Cacheable 로 캐시 저장할 때와 반환 값이 달라서 redisTemplate 사용해 캐시 업데이트
+		redisTemplate.opsForValue().set("calendar::" + request.getUserId() + ":" + request.getId(), calendar);
```
entity를 그대로 캐싱하시면 안됩니다
lazy loading 문제가 발생하거나 영속성 컨텍스트의 관리범위가 늘어나기 때문에 문제가 생길 가능성이 있습니다. 가능하면 dto 형태로 저장해주셔야합니다.

jooyeon0907 replied:
아아 네 알겠습니다 !  기존 사용 중인 dto 인 CalendarDto.Response 로 저장해도 괜찮겠죠??

CommonService 에서 
```

@Cacheable(value = "calendar", key = "#userId + ':' + #calendarId")
	public Calendar getCalendar(Long userId, Long calendarId) {
	
```
이 부분도  반환 타입을 CalendarDto.Response 로 해야겠네요

jooyeon0907 replied:
흠 CalendarDto.Response 는 API 응답용으로 설계된 것인데 의미가 혼동될 것 같아서
CalendarDto 에 캐시 전용 DTO 를 추가해서 해당 dto 로 캐싱하도록 하겠습니다!

mingming-mentor commented on src/main/java/com/zerobase/timate/service/CalendarService.java:
```diff
@@ -28,6 +27,8 @@ public class CalendarService {
+	private final RedisTemplate<String, Object> redisTemplate;
+
 	@Transactional
 	public CalendarDto.Response create(Request request) {
```
create시에는 캐싱을 따로 안하나요?

jooyeon0907 replied:
create 시에 캐싱을 할 생각은 못 했습니다..!
캘린더 저장 성공 후, getCalendar() 호출로 캐싱을 하면 된다는 말씀 맞을까요 ??

mingming-mentor replied:
네 맞습니다. create한 결과물을 return하면 @Cacheable로도 캐싱 가능합니다

mingming-mentor reviewed (approved):
수고 많으셨습니다

jooyeon0907 committed:
feat: 피드백 반영 하여 코드 수정
- 캐싱을 위한 CalendarDto 추가하여 기존 엔티티를 캐싱하던 코드를 dto 형태로 캐싱하도록 수정
  - 엔티티를 그대로 캐싱하게되면 lazy loading 또는 영속성 컨텍스트의 관리 범위가 늘어나는 문제가 생길 수 있음
- 캐시 관리를 어노테이션으로 통일 (jsonRedisTemplate 제거하고, @CachePut 사용)
- 캘린더 저장 성공 후 Calendar 캐싱 하기

jooyeon0907 committed:
test: Schedule 캐싱 테스트 추가

jooyeon0907 commented:
@mingming-mentor 님 피드백 반영하여 올렸는데 확인 한번 해 주시면 감사하겠습니다!

mingming-mentor commented:
@jooyeon0907 님~ 금요일에 제가 멘토링 끝난 이후에 메세지를 주셔서 제가 확인을 못했어요. 지금 피드백드리는 기간은 끝났지만 피드백드렸습니다. 수고 많으셨어요
